### PR TITLE
Add lowering for logcumsumexp

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118753
* #119809

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
sgtm. Can you add a test in `test/inductor/test_torchinductor.py`? Can you also put up some benchmarks?

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -4996,6 +4996,25 @@ def cumprod(x, axis=None, dtype=None):
...
+def logcumsumexp(x, dim):
+    if is_integer_dtype(x.get_dtype()) or is_boolean_dtype(x.get_dtype()):
+        dtype = torch.int64
+    else:
+        dtype = x.get_dtype()
+
+    def log_add_exp(a, b):
+        min_v = ops.minimum(a, b)
+        max_v = ops.maximum(a, b)
+        return ops.log1p(ops.exp(min_v - max_v)) + max_v
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:5009]: We probably need this special case for infinities:
https://github.com/pytorch/pytorch/blob/1bf9ddf130a15abc7cbbe905baf138fb1477b9fe/aten/src/ATen/native/cuda/LogcumsumexpKernel.cu#L38-L44

File: torch/_decomp/__init__.py [hunk 2]
```diff
@@ -334,6 +334,7 @@ def core_aten_decompositions() -> Dict[torch._ops.OperatorBase, Callable]:
             aten.linspace,
             aten.logaddexp,
             aten.logaddexp2,
+            aten.logcumsumexp,
```
Comment (peterbell10) [inline torch/_decomp/__init__.py:337]: I presume this is from a previous version where you added a decomp?

**Review by peterbell10 — CHANGES_REQUESTED** [review]

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -4996,6 +4996,27 @@ def cumprod(x, axis=None, dtype=None):
     return result
 
 
+@register_lowering(aten.logcumsumexp)
+def logcumsumexp(x, dim):
+    def log_add_exp_helper(a, b):
+        min_v = ops.minimum(a, b)
+        max_v = ops.maximum(a, b)
+        if min_v != max_v or ops.isfinite(min):
```
Comment (lezcano) [inline torch/_inductor/lowering.py:5004]: Use `ops.where` and `ops.*` handlers everywhre (ne, ops.or...)

Comment (amjames) [reply torch/_inductor/lowering.py:5004]: I did catch this with some help from peter. Interesting to me that the `!=` is not handled by operator overloading the way `+` and `-` are on the next line for example. Is there a reason that logical operators are not overloaded in this way or is that an oversight that could be corrected?

Comment (peterbell10) [reply torch/_inductor/lowering.py:5004]: btw here is where the operators are implemented:
https://github.com/pytorch/pytorch/blob/b73a2b779586b23e62759cc41b3b0724bbfa3080/torch/_inductor/virtualized.py#L802-L803

**Review by lezcano — COMMENTED** [review]
SG, but you uncovered a tiny bug. @peterbell10 's implementation is a bit leaky. You'll need to unwrap the value within the OpsValue at some point. Touch base with him to figure out where is best to do this. Otherwise you can simply use `ops.` handlers.

**Comment by lezcano** [comment]
Ah, I see the errors come from the previuous PR. In any case, it looks like this should fix a few bugs in the lowerings, I guess...

File: torch/_inductor/lowering.py [hunk 4]
```diff
@@ -5062,6 +5062,26 @@ def cumprod(x, axis=None, dtype=None):
...
+    def log_add_exp_helper(a, b):
+        min_v = ops.minimum(a, b)
+        max_v = ops.maximum(a, b)
+        mask = (min_v != max_v) | (~ops.isinf(min_v))
+        return ops.where(mask, ops.log1p(ops.exp(min_v - max_v)) + max_v, a)
+
+    dtype = x.get_dtype()
+    if len(x.get_size()) == 0:
+        assert dim in [0, -1]
+        return to_dtype(x, dtype, copy=True)
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:5076]: NIT:
```suggestion
        return clone(x)
```

File: test/inductor/test_torchinductor_opinfo.py [hunk 5]
```diff
@@ -322,6 +322,7 @@ def wrapper_noop_set_seed(op, *args, **kwargs):
     ("cauchy", "cuda"): {"reference_in_float": True},
     ("cummax", "cuda", f16): {"atol": 5e-4, "rtol": 0.002},
     ("cumprod", "cuda"): {"reference_in_float": True, "atol": 7e-5, "rtol": 0.002},
+    ("logcumsumexp", "cuda"): {"atol": 8e-4, "rtol": 0.001},
```
Comment (peterbell10) [inline test/inductor/test_torchinductor_opinfo.py:325]: This tolerance is rather high but I see it's only relevant to the backward case. Do you think we could ad a `grad_atol` and `grad_rtol` to the check function so we don't have to be so loose for the forward case?

**Review by peterbell10 — APPROVED** [review]

**Comment by jeffdaily** [comment]
This PR appears to have broken ROCm on the trunk.

[test/inductor/test_torchinductor_codegen_dynamic_shapes.py::DynamicShapesCodegenGPUTests::test_logcumsumexp_dynamic_shapes_cuda](https://hud.pytorch.org/failure?name=rocm%20%2F%20linux-focal-rocm5.7-py3.8%20%2F%20test%20(default%2C%203%2C%206%2C%20linux.rocm.gpu.2)&jobName=undefined&failureCaptures=%5B%22inductor%2Ftest_torchinductor_codegen_dynamic_shapes.py%3A%3ADynamicShapesCodegenGPUTests%3A%3Atest_logcumsumexp_dynamic_shapes_cuda%22%5D)

**Comment by jeffdaily** [comment]
Can this be conditionally lowered not for ROCm?  How do we handle other cases where ROCm's triton support is still catching up?  CC @jataylo

**Comment by peterbell10** [comment]
> Can this be conditionally lowered not for ROCm? 

It does fallback on ROCm which is why the test failed. We can either add a skip and merge now, or wait for #119369.

**Comment by jataylo** [comment]
@jeffdaily @peterbell10 merging the ir.Scan enablement PR for ROCm now https://github.com/pytorch/pytorch/pull/119369 if this would allow this PR to work for us too.

**Pull request closed without merging**